### PR TITLE
320 rounds testing: migration and reload test

### PR DESCRIPTION
**Pull request by nicholasguoalgorand**
## Summary

Write a test that initialized ledger + db, pushes txns through the network, then restart ledger + migrate db. We want to ensure that all the online accounts data is correct after this process.

## Test Plan

This is a test

### CODE REVIEW

File: ledger/ledger_test.go [hunk 1]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
 	}
 }
 
+// TestLedgerMigrateV6ShrinkDeltas checks the ledger has correct account state
+// after performing a db migration from V6
+func TestLedgerMigrateV6ShrinkDeltas(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	accountDBVersion = 6
```
Comment (algorandskiy) [inline ledger/ledger_test.go:1952]: defer `accountDBVersion=7` in case of the test failure just in case

File: ledger/ledger_test.go [hunk 2]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
 	}
 }
 
+// TestLedgerMigrateV6ShrinkDeltas checks the ledger has correct account state
+// after performing a db migration from V6
+func TestLedgerMigrateV6ShrinkDeltas(t *testing.T) {
```
Comment (algorandskiy) [inline ledger/ledger_test.go:1949]: please debrief the test scenario

File: ledger/ledger_test.go [hunk 3]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+	}()
+	genesisInitState, initKeys := ledgertesting.GenerateInitState(t, testProtocolVersion, 10_000_000_000)
+	const inMem = false
+	cfg := config.GetDefaultLocal()
+	cfg.MaxAcctLookback = proto.MaxBalLookback
+	log := logging.TestingLog(t)
+	log.SetLevel(logging.Info) // prevent spamming with ledger.AddValidatedBlock debug message
+	dbs, _, err := openLedgerDB(dbName, inMem)
+	require.NoError(t, err)
+	err = dbs.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
```
Comment (algorandskiy) [inline ledger/ledger_test.go:1969]: comment why we are creating them

File: ledger/ledger_test.go [hunk 4]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+			txHeader := transactions.Header{
+				Sender:      addresses[curAddressIdx],
+				Fee:         basics.MicroAlgos{Raw: proto.MinTxnFee * uint64(feeMult)},
+				FirstValid:  latest + 1,
+				LastValid:   latest + maxValidity,
+				GenesisID:   t.Name(),
+				GenesisHash: crypto.Hash([]byte(t.Name())),
+			}
+
+			correctPayFields := transactions.PaymentTxnFields{
```
Comment (algorandskiy) [inline ledger/ledger_test.go:2022]: please mix-in keyreg transactions flipping account state as well

File: ledger/ledger_test.go [hunk 5]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+	sort.SliceStable(addresses, func(i, j int) bool { return bytes.Compare(addresses[i][:], addresses[j][:]) == -1 })
+
+	onlineTotals := make([]basics.MicroAlgos, maxBlocks+1)
+	curAddressIdx := 0
+	maxValidity := basics.Round(20) // some number different from number of txns in blocks
+	txnIDs := make(map[basics.Round]map[transactions.Txid]struct{})
+	// run for maxBlocks rounds with random payment transactions
+	// generate 1000 txn per block
+	for i := 0; i < maxBlocks; i++ {
+		stxns := make([]transactions.SignedTxn, 10)
```
Comment (algorandskiy) [inline ledger/ledger_test.go:2006]: add a random number of transactions say from 7 to 15

File: ledger/ledger_test.go [hunk 6]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+		require.Equal(t, origRewardsBalances[i], acct.MicroAlgos)
+		require.Equal(t, origBalances[i], wo)
+
+		oad, err := l.LookupAgreement(balancesRound, addr)
+		require.NoError(t, err)
+		require.Equal(t, origAgreementBalances[i], oad.MicroAlgosWithRewards)
+
+		// TODO:
+		// add a test checking all committed pre-reload entries are gone
+		// add as a tracker test
```
Comment (algorandskiy) [inline ledger/ledger_test.go:2123]: remove todo

File: ledger/ledger_test.go [hunk 7]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+	}
+
+	shorterLookback := config.GetDefaultLocal().MaxAcctLookback
+	require.Less(t, shorterLookback, cfg.MaxAcctLookback)
+	cfg.MaxAcctLookback = shorterLookback
+	l.cfg = cfg
+	accountDBVersion = 7
+	err = dbs.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		tx.ExecContext(ctx, "DROP TABLE IF EXISTS onlineaccounts")
+		//tx.ExecContext(ctx, "DROP TABLE IF EXISTS txtail")
```
Comment (algorandskiy) [inline ledger/ledger_test.go:2089]: why commented out?

File: ledger/ledger_test.go [hunk 8]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+			require.NoError(t, l.CheckDup(proto, nextRound, i-maxValidity, i, txid, ledgercore.Txlease{}))
+		}
+	}
+
+	// check an error latest-1
+	for txid := range txnIDs[latest-1] {
+		require.Error(t, l.CheckDup(proto, nextRound, latest-maxValidity, latest-1, txid, ledgercore.Txlease{}))
+	}
+
+	shorterLookback := config.GetDefaultLocal().MaxAcctLookback
```
Comment (algorandskiy) [inline ledger/ledger_test.go:2082]: l.Close()

File: ledger/ledger_test.go [hunk 9]
```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+	accountDBVersion = 7
+	err = dbs.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		tx.ExecContext(ctx, "DROP TABLE IF EXISTS onlineaccounts")
+		//tx.ExecContext(ctx, "DROP TABLE IF EXISTS txtail")
+		tx.ExecContext(ctx, "DROP TABLE IF EXISTS onlineroundparamstail")
+		return nil
+	})
+	l.genesisProtoVersion = protocol.ConsensusCurrentVersion
+	l.genesisProto = config.Consensus[protocol.ConsensusCurrentVersion]
+	l.reloadLedger()
```
Comment (algorandskiy) [inline ledger/ledger_test.go:2095]: l = OpenLedger()

File: ledger/ledger_test.go (hunk 7 above)
Comment (nicholasguoalgorand) [reply ledger/ledger_test.go:2089]: Because if we delete txtail, test that makes sure item that is in txtail fails

Comment (algorandskiy) [reply ledger/ledger_test.go:2089]: Interesting, in theory it should migrate data correctly and init in-memory data while replaying blocks. Looks like worth debugging

File: ledger/ledger_test.go (hunk 5 above)
Comment (nicholasguoalgorand) [reply ledger/ledger_test.go:2006]: so random # btwen 7-15 instead of 10 each time?

Comment (algorandskiy) [reply ledger/ledger_test.go:2006]: yes, please

File: ledger/ledger_test.go [hunk 10]
```diff
@@ -1943,6 +1944,213 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+		accountsCreateOnlineAccountsTable(ctx, tx)
+		accountsCreateTxTailTable(ctx, tx)
+		accountsCreateOnlineRoundParamsTable(ctx, tx)
+		return nil
+	})
+	l, err := OpenLedger(log, dbName, inMem, genesisInitState, cfg)
+	require.NoError(t, err)
+	defer func() {
+		l.Close()
+		os.Remove(dbName + ".block.sqlite")
```
Comment (algorandskiy) [inline ledger/ledger_test.go:1983]: remove -wal, -shm

Comment (algorandskiy) [inline ledger/ledger_test.go:1987]: remove -wal, -shm

File: ledger/ledger_test.go [hunk 11]
```diff
@@ -1943,6 +1944,213 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
...
+	for txid := range txnIDs[latest-1] {
+		require.Error(t, l.CheckDup(proto, nextRound, latest-maxValidity, latest-1, txid, ledgercore.Txlease{}))
+	}
+
+	shorterLookback := config.GetDefaultLocal().MaxAcctLookback
+	require.Less(t, shorterLookback, cfg.MaxAcctLookback)
+	cfg.MaxAcctLookback = shorterLookback
+	l.cfg = cfg
+	accountDBVersion = 7
+	l.Close()
```
Comment (algorandskiy) [inline ledger/ledger_test.go:2103]: move `l.Close()` up - to above changing the config

Comment (algorandskiy) [inline ledger/ledger_test.go:2103]: move `l.Close()` up - to above changing the config

**Review by algorandskiy — COMMENTED** [review]
LGTM - couple comments and gofmt complaint

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**